feat(ingest): add ingestion source for SAP Analytics Cloud

Masterchen09 opened the pull request:
@hsheth2 As discussed, here is the ingestion source for the SAP Analytics Cloud with everything included (docs, tests, configuration in the UI, ...).

Two remarks regarding two changes which are also included in this PR, but have nothing to do with the SAP Analytics Cloud source:

- It seems the [getLogoFromPlatform](https://github.com/datahub-project/datahub/compare/master...Masterchen09:feat-sap-analytics-cloud?expand=1#diff-3fcfec2e1275f28e12793b1656059382719fcc05ad7d577a2810f1e712dcbced) function is not used anymore, therefore I have removed it...
- I have noticed that the ingestion sources are currently not sorted when they are selected in the UI...therefore I have added a sort function in [SelectTemplateStep](https://github.com/datahub-project/datahub/compare/master...Masterchen09:feat-sap-analytics-cloud?expand=1#diff-1185d29088749d7201b3c1d7fc84b7d3e72cacb4bf5264451abae123fef731d5) to sort the sources alphabetically, while making sure, that the "Custom" source is always displayed last...;-)

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced integration with SAP Analytics Cloud (SAC) for data ingestion.
  - Support for various SAC connection parameters, including tenant URL and token URL.
  - New source configuration for SAC integration introduced, expanding platform capabilities.

- **Bug Fixes**
  - Improved sorting logic for displayed sources in the UI to prioritize non-custom entries.

- **Documentation**
  - New documentation detailing SAC integration setup, configuration notes, and limitations.

- **Tests**
  - Introduced integration tests for the SAC API to ensure proper interaction and validation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hsheth2 commented on metadata-ingestion/src/datahub/utilities/logging_manager.py:
```diff
@@ -278,3 +278,6 @@ def configure_logging(debug: bool, log_file: Optional[str] = None) -> Iterator[N
 logging.getLogger("snowflake").setLevel(level=logging.WARNING)
 # logging.getLogger("botocore").setLevel(logging.INFO)
 # logging.getLogger("google").setLevel(logging.INFO)
+logging.getLogger("pyodata.client").setLevel(logging.WARNING)
```
loggers have hierarchy, so it think this should work

```suggestion
logging.getLogger("pyodata").setLevel(logging.WARNING)
```

hsheth2 commented on metadata-ingestion/setup.py:
```diff
@@ -471,6 +477,7 @@
     "fivetran": snowflake_common | bigquery_common | sqlglot_lib,
     "qlik-sense": sqlglot_lib | {"requests", "websocket-client"},
     "sigma": sqlglot_lib | {"requests"},
+    "sac": sac,
```
imo sac isn't a sufficiently well-know abbreviation - let's name the source `sap-analytics-cloud` or `sap-sac`

Masterchen09 replied:
At least in context of SAP, SAC ist a common abbreviation for SAP Analytics Cloud...if I search for SAC in Google or Bing in an incognito window, besides the Swiss Alpine Club (😂), the next results are for the SAP Analytics Cloud (but of course there is still some kind of geo localization and the results are probably specific for Germany).

-> sap-sac is definitely wrong (the S in SAC stands for SAP) and sap-analytics-cloud is too long...

As the abbreviation would be also used for the platform urn, I would like to stick with SAC...;-)

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/common/subtypes.py:
```diff
@@ -71,3 +74,7 @@ class BIAssetSubTypes(str, Enum):
+
+    # SAP Analytics Cloud
+    SAC_STORY = "Story"
+    SAC_APPLICATION = "Application"
```
nice

hsheth2 commented on metadata-ingestion/tests/integration/sac/test_sac.py:
```diff
@@ -0,0 +1,308 @@
+
+        requests_mock.get(f"{MOCK_TENANT_URL}/api/v1/$metadata", content=match_metadata)
+
+    def match_resources(request, context):
```
would like to avoid nested functions if possible

hsheth2 commented on metadata-ingestion/docs/sources/sac/sac_pre.md:
````diff
@@ -0,0 +1,45 @@
+        env: PROD
+```
+
+The key in the connection mapping is the technical name of the connection resp. its id.
````
bit confused by this line

Masterchen09 replied:
I have changed it a bit, not sure if it is less confusing then...

You have to maintain the connections to the different systems in the SAC...there you have to specify a name: https://developers.sap.com/tutorials/hana-cloud-analytics-cloud-consume-2/jcr:content.github-proxy.1633841405.file/step7-I7.png

The key in the connection mapping dictionary is this name and is used to find the connection in the dictionary...

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+                )
+            else:
+                test_report.internal_failure = True
+                test_report.internal_failure_reason = f"{e}"
```
when might this happen?

Masterchen09 replied:
Never ;-)

-> Fixed

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+                    )
+
+                    yield MetadataWorkUnit(
+                        id=f"dashboard-browse-paths-{dashboard_urn}", mcp=mcp
```
`yield mcp.as_workunit()`

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+                        changeType=ChangeTypeClass.UPSERT,
+                        entityUrn=dashboard_urn,
+                        aspectName="browsePaths",
+                        aspect=BrowsePathsClass(
```
you only need to pass entityUrn and aspect - the rest will be auto-populated

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+        if self.config.ingest_stories or self.config.ingest_applications:
+            resources = self.get_resources()
+
+            for resource in resources:
```
it'd be good to pull some of these out into helper methods - the `get_workunits_internal` method is a bit long and hence hard to follow

Masterchen09 replied:
Should be better now?

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+                elif column.data_type == "int32":
+                    native_data_type = f"{column.data_type}({column.precision})"
+                elif column.max_length is not None:
+                    native_data_type = f"{column.data_type}({column.max_length})"
```
would be nicer to move this into a helper method called `get_schema_field_native_data_type`

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+                    platform_instance = connection.platform_instance
+                    env = connection.env
+                else:
+                    platform = model.system_type.lower()
```
does it make sense to print some debug logs here? e.g. no connection_mapping found, doing best-effort inference

or are the logs below sufficient?

Masterchen09 replied:
I have added a log with info level (I think debug is not enough?) - besides the log message you would notice that the best-effort approach was used, because there will be platform instances which are named after the connections in SAC...

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+            logger.warning(
+                f"Unknown system type {model.system_type} for model with id {model.namespace}:{model.model_id} and external id {model.external_id}"
+            )
+            self.report.report_warning(
+                "unknown-system-type",
+                f"Unknown system type {model.system_type} for model with id {model.namespace}:{model.model_id} and external id {model.external_id}",
+            )
```
`.warning` will also automatically log the message, so you can remove the logger.warning call

```suggestion
            self.report.warning(
                title="Cannot generate upstream lineage",
                message="Unknown system type - lineage will be incomplete.",
                context=f"system type {model.system_type} for model with id {model.namespace}:{model.model_id} and external id {model.external_id}",
            )
```

Masterchen09 replied:
There were probably some API changes in the past I missed (i.e. #10828), but also omiting everything except entityUrn and aspect in MetadataChangeProposalWrapper or the mcp.as_workunit method...;-)

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+            elif data_type in ("decimal", "int32"):
+                return SchemaFieldDataTypeClass(type=NumberTypeClass())
+            else:
+                logger.warning(f"Unknown data type {data_type} found")
```
```suggestion
                logger.info(f"Unknown data type {data_type} found")
```

Masterchen09 replied:
I have replaced the log with report_warning -> I think it is better to have this as a warning...usually there are not many possibilities for the data types in import data models - in case there will be ever an unknown data type, we should know it to enhance the method accordingly...I am affraid a log with info level will be missed easily...

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+        if property_type == "DATE":
+            return SchemaFieldDataTypeClass(type=DateTypeClass())
+        else:
+            if data_type == "string":
```
so property_type is capitalized and data_type is lower case?

Masterchen09 replied:
Yes, we shouldn't really question things like this...SAP and REST APIs is a "complicated" topic (at least in context of SAC)...

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -0,0 +1,808 @@
+                grant_type="client_credentials",
+            )
+
+            session.register_compliance_hook(
```
is there a way to extract more of this connection logic out into a helper? don't like that it's repeated across init and test_connection

e.g. `self.session, self.client = make_sac_connection(config)`

hsheth2 reviewed (commented):
Mainly focused on the python code in this pass - this is looking pretty good

Masterchen09 commented:
@hsheth2 I have noticed, that the ingestion of the schema metadata of Import Data Models is extremely slow (20 seconds vs. 15 minutes with our productive SAC tenant), because the Import Data Service of the SAC is extremely slow...therefore I have added a config (by default it is enabled) to be able to disable the ingestion of schema metadata.

Additionally the Import Data Service of the SAC is frequently returning an 500 Internal Server Error...therefore I have also added a HTTPAdapter with a Retry to the Session object...ingesting the schema metadata is relatively new (I have implemented it before creating the PR, while we are using the rest of the source for some time): In the last two weeks the Internal Server Error already occurred three times, not sure if the Retry will help to solve this issue, but we will see...

hsheth2 commented on metadata-ingestion/src/datahub/ingestion/source/sac/sac.py:
```diff
@@ -359,73 +274,110 @@ def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
+                aspect=BrowsePathsV2Class(
+                    path=[
+                        BrowsePathEntryClass(id=folder_name)
+                        for folder_name in resource.ancestor_path.split("/")
```
is there any danger of folder names having `/` in them?

Masterchen09 replied:
```/``` are replaced with ```%2F```, when the folders which are returned by the API are evaluated:

https://github.com/datahub-project/datahub/blob/74e0c7f74d5152468f12f97390787daf18bd4e97/metadata-ingestion/src/datahub/ingestion/source/sac/sac.py#L620-L623

The replacement was originally in there because of the browse paths v1, which are still ingested along the browse paths v2 - do we need the v1 browse paths anymore or should I remove them? - ...in the old browse paths the slashes were used to separate the levels, I don't think the replacement is needed for v2 anymore...? However for the config.folder_pattern it's probably easier if the folders are joined using the slash inkl. the replacement of slashes in folder names...:/

hsheth2 replied:
We don't really need v1 browse paths anymore, but I agree that having this around makes `folder_pattern` simpler so it's fine to leave it in

hsheth2 commented on datahub-web-react/src/app/ingest/source/builder/SelectTemplateStep.tsx:
```diff
@@ -104,6 +104,18 @@ export const SelectTemplateStep = ({ state, updateState, goTo, cancel, ingestion
+
+        if (b.name === 'custom') {
+            return -1;
+        }
```
this is just to push the "custom" button to the bottom, right?

Masterchen09 replied:
Yes, exactly

hsheth2 commented:
@Masterchen09 looks like the cypress tests that are failing are `mutations/ingestion_source.js` and `mutations/managing_secrets.js`, which worries me a bit given the changes here. Not sure if it's a flaky test or a real failure

Masterchen09 commented:
@hsheth2 Thanks for mentioning it, I didn't noticed that some tests are failing.

For ```mutations/ingestion_source.js``` it sounds logical, that the test is failing: Because of the sorting Snowflake is not visible anymore without scrolling...I hope this is fixed now.

I have currently no idea regarding ```mutations/managing_secrets.js```- the changes to the UI should have no influence on this...let's see whether the test is passing now or not...

Masterchen09 commented:
@hsheth2 The fix for ```mutations/ingestion_source.js``` yesterday solved the issue...the ```mutations/managing_secrets.js``` failed again, but I think the error message is just a bit misleading:

```AssertionError: Timed out retrying after 10000ms: Expected to find content: 'Successfully created Secret!' but never did.```

This is not the reason for the first attempt to fail, only the reason for the second and third attempt...the error message is caused by the failing first attempt, because the secret already exists and will not be created in the second and third attempt - the reason for failing the first attempt is the same reason why ```mutations/ingestion_source.js``` failed: Snowflake is not visible without scrolling...I have applied the same fix to ```mutations/managing_secrets.js``` which I have added yesterday to ```mutations/ingestion_source.js```...let's see whether the test is successful now.

There is also another fix for the frozen dataclasses...from a functional point view it makes no difference, but is not correct to have a set in a frozen dataclass - now it's a frozenset. ;-)

hsheth2 commented:
CI failures are unrelated (nifi, nested domains), so merging through.
